README


     Eventually Persistent Ram Disk

EPRD is an eventually persistent ramdisk.

This blockdevice can be configured in a number of ways.
The first way to configure EPRD is to assign a certain amount
of ram to eprd. In this mode no disk io will ever take 
place except when the device is configured or unconfigured.
When the device is created it will simply load the data
from disk in memory. When the device is remove the data will
be save to disk.

You can create this type of disk backed ramdisk like this:
insmod ./eprd.ko
./eprd_setup -f /data/persramdisk.img -s 1G -c
Now device /dev/eprda will be created.

To unload the device simply type :
rmmod eprd or eprd_setup -d /dev/eprda

** Please note that the -c parameter means create and should
only be used the first time. **

You can however also create a ramdisk that will have it's
changes flushed to disk every N seconds and instead of keeping
everything in ram it only allocated 512M for caching.

./eprd_setup -f /data/saverimg -s 10G -m 3 -c -p512M

In this example the data is flushed to disk every 3 seconds.

Although this is already somewhat safer there is an even
more 'reliable' mode of operation.

./eprd_setup -f /data/saverimg -s 10G -m 3 -c -p512M -b

EPRD supports barriers. So when the device is configured with
the -b option a sync() on the filesystem will mean a sync of 
the blockdevice and all 'dirty' blocks in the cache will be flushed
to disk.

You can easily test the effect of barrier support when you format
eprd with ext4 or xfs. In this example I will use ext4
mkfs.ext4 /dev/eprda
mount /dev/eprda /mnt -obarrier=0

root@saturn:/mnt# writetest
running: 1000 ops
Elapsed time: 0.029147 seconds
1000 ops: 34308.85 ops per second

While after : mount -oremount,barrier=1 /mnt
root@saturn:/mnt# writetest
running: 1000 ops
Elapsed time: 42.926026 seconds
1000 ops:   23.30 ops per second

As you can see the difference is huge.

EPRD can also be used with raw devices:
./eprd_setup -f /dev/sdc -m 3 -p512M -b

The example above creates a new eprd device that is backed by
a raw device. Please note that you do not need to specify -c
when a raw device is used.

tools/writetest is a small tool provided by Oracle
for testing Berkeley DB performance. 


Changelog
release                         : 0.4.2  2012-05-30
                                : Close the filedescriptor when deregistering the
                                : eprd device.
release                         : 0.4.1  2012-05-16
                                : Small fix of determine_iotype.
                                : Another small fix that allows compilation 
                                : on 32 bit systems
release                         : 0.4.0  2012-05-08
                                : Improve the caching algorithm when smaller then 4k
                                : blocks are written or read.
release                         : 0.3.1  2012-05-04
                                : Fix a possible deadlock
release                         : 0.3.0  2012-05-03
                                : Added a quick fix for a huge data corruption bug!
                                : This bug was introduced when I added support for
                                : smaller then 4k sector sizes.
release                         : 0.2.0  2012-04-27
                                : Make eprdcontrol minor dynamic
                                : as suggested by Roman Prokopovich. Thanks!
release                         : 0.1.7  2012-04-16
                                : EPRD now recognizes RANDOM or SEQUENTIAL IO.
                                : Only when the io is of random nature EPRD will
                                : cache. Otherwise it will just write the 
                                : incoming data to disk without delay.
release                         : 0.1.6  2012-04-15
                                : EPRD now supports other then 4k sector sizes
                                : and now defaults to 512 bytes since for example
                                : vmware can not handle larger sector size.
                                : The caching algorithm is much improved as well.
release                         : 0.1.2  2012-04-08
        Resolved problems       : Resolved a data corruption bug
                                : caused by insufficient locking.
                                : ext4 and xfs now log that barriers are
                                : unavailable when eprd has barriers disabled.
                                : Solves a problem where eprd would still
                                : write to disk when it should only write when
                                : being registered or unregistered.
                                : Improved caching algorithm.
        Known problems          :
release                         : 0.1.1  2012-04-06
        Resolved problems       : Added _LARGEFILE64_SOURCE to eprd_setup
                                : to make it possible to use large devices.
        Resolved problems       : Added _LARGEFILE64_SOURCE to eprd_setup
                                : to make it possible to use large devices.
        Known problems          :

Initial release                 : 0.1.0  2012-04-04
        Resolved problems       :
        Known problems          :


test.sh
#logrotate -f /etc/logrotate.conf
insmod ./eprd.ko
sleep 1
#./eprd_setup -f /data/eprddta.img -m 3 
#./eprd_setup -f /data/eprddta
#./eprd_setup -f /data/eprddta -p 1000M -m 5 -b
#./eprd_setup -f /data/eprddta.img -c -s 8192M -p 512M -m 3 -b 
#./eprd_setup -f /dev/mapper/vg_dynamic11-lv_data -m3 -p 4096M
#sleep 2
#mkfs.xfs -f /dev/eprda -b size=4096
#mount /dev/eprda /mnt -o nobarrier 
#mount /dev/eprda /mnt 
#mkfs.ext4 /dev/eprda
#mount /dev/eprda /mnt -o barrier=0
#mount /dev/eprda /mnt -o barrier=1


eprd.h
#define BLKSIZE 4096
#define EPRD_NAME_SIZE     64  /* Max lenght of the filenames*/
#define EPRD_SET_DTAFD     0xFE00
#define EPRD_SET_DEVSZ     0xFE03
#define EPRD_REGISTER      0xFE04
#define EPRD_DEREGISTER    0xFE05
#define EPRD_SET_COMMIT_INTERVAL 0xFE06
#define EPRD_INIT          0xFE07
#define EPRD_BARRIER       0xFE08
#define EPRD_CACHESIZE     0xFE09
#define EPRD_SET_SECTORSIZE  0xFE0A

#define RANDOM 0x01
#define SEQUENTIAL 0x02
#define KERNEL_SECTORSIZE 512

#ifdef __KERNEL__
#include <linux/bio.h>
#include <linux/blkdev.h>
#include <linux/spinlock.h>
#include <linux/mutex.h>
#include <linux/rbtree.h>
#include <linux/file.h>

#define EPRDCRIT(f, arg...) \
        printk(KERN_CRIT "EPRD : " f "\n", ## arg)
#define EPRDERR(f, arg...) \
        printk(KERN_NOTICE "EPRD : " f "\n", ## arg)
#define EPRDINFO(f, arg...) \
        printk(KERN_INFO "EPRD : " f "\n", ## arg)

struct rbtree {
        spinlock_t rbtree_lock;
        struct rb_root rbtree_root;
};

typedef struct {
    struct file *fp;
    mm_segment_t fs;
} file_info_t;

struct rbdata {
   struct rb_node node;
   unsigned long blocknr;
   char data[BLKSIZE];
   u_char valid;
   int dirty;
   unsigned int hitcount;
   unsigned int pass;
};

struct eprd_device {
        struct list_head list;
        int                     major_num;
        int                     eprd_device_number;
        int                     active;
        int                     (*ioctl)(struct eprd_device *, int cmd, 
                                     unsigned long arg); 
        unsigned long           nsectors;
        int                     logical_block_size;
        struct file *           eprd_dta_file;
        struct block_device     *eprd_device;
        struct mutex            eprd_ctl_mutex;
        unsigned long           size;
        spinlock_t              lock;
        struct gendisk          *gd;
        struct rbtree           rb_tree;
        struct workqueue_struct *wqueue;
        struct task_struct      *eprd_thread;
        struct bio_list         eprd_bio_list;
        struct request_queue    *rqueue;
        char                    *devname;
        atomic_t                flush; 
        atomic_t                wcachelock; 
        atomic_t                commit; 
        unsigned int            commit_interval;
        int                     barrier;
        int                     stop;
/*Holds the type of IO random or sequential*/
        int                     iotype;
/*Last blocknr written or read*/
        unsigned long           lastblocknr;  
/*Incremented if current blocknr == lastblocknr -1 or +1 */
        unsigned int            insequence;  
        unsigned long           cachesize;
        unsigned long           cacheentries;
        struct mutex            cachelock;
        wait_queue_head_t       eprd_event;
        wait_queue_head_t       cache_event;
        struct timer_list       flush_timer;
};

typedef struct {
        struct work_struct work;
        struct eprd_device *device;
} cache_worker_t;

static loff_t eprd_get_size(loff_t , loff_t , struct file *);
static int eprd_file_write(struct file *, u8 *, const int, loff_t);
int eprd_sync(struct eprd_device *);
int is_continues(u_char, int *, int *);

#endif


eprd_setup.c
#define _GNU_SOURCE
#define _LARGEFILE64_SOURCE

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <stdarg.h>
#include <errno.h>

#include "eprd.h"

#define die_ioctlerr(f...) { fprintf(stderr,(f)); flock(fd, LOCK_UN); exit(-1); }
#define die_syserr() { fprintf(stderr,"Fatal system error : %s",strerror(errno)); exit(-2); }

int errno;

struct option_info {
	int create;
	int sectorsize;
	unsigned long cachesize;
	unsigned int commit_interval;
	char *datafile;
	off64_t dtasize;
	char *device;
	int barrier;
};

void *s_malloc(size_t size)
{
	void *retval;
	retval = malloc(size);
	if (!retval)
		die_syserr();
	return retval;
}

void *s_realloc(void *ptr, size_t size)
{
	void *retval;
	retval = realloc(ptr, size);

	if (!retval)
		die_syserr();
	return retval;
}

void *as_sprintf(const char *fmt, ...)
{
	/* Guess we need no more than 100 bytes. */
	int n, size = 100;
	void *p;
	va_list ap;
	p = s_malloc(size);
	while (1) {
		/* Try to print in the allocated space. */
		va_start(ap, fmt);
		n = vsnprintf(p, size, fmt, ap);
		va_end(ap);
		/* If that worked, return the string. */
		if (n > -1 && n < size)
			return p;
		/* Else try again with more space. */
		if (n > -1)	/* glibc 2.1 */
			size = n + 1;	/* precisely what is needed */
		else		/* glibc 2.0 */
			size *= 2;	/* twice the old size */
		p = s_realloc(p, size);
	}
}

static struct option_info mkoptions;

int eprd_setup(int op, int fd)
{
	int ffd, i;
	char *pass;
	char *filename;
	struct stat stbuf;
	unsigned long fsize;
	int mode;
	int ret = 0;

	switch (op) {
	case EPRD_SET_DTAFD:
		if (mkoptions.create)
			mode = O_CREAT | O_RDWR;
		else
			mode = O_RDWR;
		if ((ffd = open(mkoptions.datafile, mode, 0600)) < 0) {
			if (ffd < 0) {
				fprintf(stderr, "Failed to open file %s",
					mkoptions.datafile);
				return 1;
			}
		}
		if (mkoptions.create)
			ret = ftruncate(ffd, mkoptions.dtasize);
		if (ret != 0) {
			fprintf(stderr, "ftruncate %s failed : %s\n",
				mkoptions.datafile, strerror(errno));
			close(ffd);
			return ret;
		}
		if (ioctl(fd, EPRD_SET_DTAFD, ffd) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_SET_DTAFD failed on /dev/eprdcontrol\n");
			close(ffd);
			return rc;
		}
		if (-1 == fstat(ffd, &stbuf)) {
			fprintf(stderr, "Failed to stat %s",
				mkoptions.datafile);
			close(ffd);
			return -1;
		}
		if (ioctl(fd, EPRD_SET_DEVSZ, mkoptions.dtasize) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_SET_DEVSZ failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_SET_COMMIT_INTERVAL:
		if (ioctl
		    (fd, EPRD_SET_COMMIT_INTERVAL,
		     mkoptions.commit_interval) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_SET_COMMIT_INTERVAL failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_CACHESIZE:
		if (ioctl(fd, EPRD_CACHESIZE, mkoptions.cachesize) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_CACHESIZE failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_SET_SECTORSIZE:
		if (ioctl(fd, EPRD_SET_SECTORSIZE, mkoptions.sectorsize) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_SET_SECTORSIZE failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_REGISTER:
		if (ioctl(fd, EPRD_REGISTER, 0) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_REGISTER failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_DEREGISTER:
		if (ioctl(fd, EPRD_DEREGISTER, (unsigned long)mkoptions.device)
		    < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_DEREGISTER failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_BARRIER:
		if (ioctl(fd, EPRD_BARRIER, (unsigned long)mkoptions.barrier) <
		    0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_BARRIER failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	case EPRD_INIT:
		if (ioctl(fd, EPRD_INIT, 0) < 0) {
			int rc = 1;
			fprintf(stderr,
				"ioctl EPRD_INIT failed on /dev/eprdcontrol\n");
			return rc;
		}
		break;
	default:
		abort();
	}
	close(ffd);
	return 0;
}

void usage(char *name)
{
	printf
	    ("Create : %s -f datafile -s size_of_blockdevice[MG] -m commit_interval [-z sectorsize(512..4096) -b(barrier) -c(create) -h(help) -p(cachesize)]\n",
	     name);
	printf("Detach : %s -d /dev/eprd_device_name\n", name);
	exit(-1);
}

int get_opts(int argc, char *argv[])
{

	int c, ret = 0;

	while ((c = getopt(argc, argv, "d:bs:hcf:m:p:z:")) != -1)
		switch (c) {
		case 'c':
			mkoptions.create = 1;
			break;
		case 'b':
			mkoptions.barrier = 1;
			break;
		case 'm':
			if (optopt == 'm')
				printf
				    ("Option -%m requires commit interval (sec) as argument.\n",
				     optopt);
			else
				sscanf(optarg, "%u",
				       &mkoptions.commit_interval);
			break;
		case 'f':
			if (optopt == 'f')
				printf
				    ("Option -%c requires a lessfs configuration file as argument.\n",
				     optopt);
			else
				mkoptions.datafile = optarg;
			break;
		case 'd':
			if (optopt == 'd')
				printf
				    ("Option -%c requires a device as argument.\n",
				     optopt);
			else {
				mkoptions.device = optarg;
			}
			break;
		case 's':
			if (optopt == 's')
				printf
				    ("Option -%c requires file size as argument.\n",
				     optopt);
			else {
				sscanf(optarg, "%lu", &mkoptions.dtasize);
				if (optarg[strlen(optarg) - 1] == 'M')
					mkoptions.dtasize *= 1024 * 1024;
				if (optarg[strlen(optarg) - 1] == 'G')
					mkoptions.dtasize *= 1024 * 1024 * 1024;
			}
			break;
		case 'p':
			if (optopt == 'p')
				printf
				    ("Option -%c requires cache size as argument.\n",
				     optopt);
			else {
				sscanf(optarg, "%lu", &mkoptions.cachesize);
				if (optarg[strlen(optarg) - 1] == 'M')
					mkoptions.cachesize *= 1024 * 1024;
				if (optarg[strlen(optarg) - 1] == 'G')
					mkoptions.cachesize *=
					    1024 * 1024 * 1024;
			}
			break;
		case 'z':
			if (optopt == 'z')
				printf
				    ("Option -%c requires sector size as argument.\n",
				     optopt);
			else {
				sscanf(optarg, "%i", &mkoptions.sectorsize);
				if (mkoptions.sectorsize > 4096)
					mkoptions.sectorsize = 0;
				if (mkoptions.sectorsize < 0)
					mkoptions.sectorsize = 0;
				if (mkoptions.sectorsize > 0) {
					ret = mkoptions.sectorsize / 512;
					ret *= 512;
					if (ret != mkoptions.sectorsize) {
						ret = -1;
						printf
						    ("The sectorsize has to be a multiple of 512 bytes\n");
					} else
						ret = 0;
				}
			}
			break;
		case 'h':
			usage(argv[0]);
			break;
		default:
			abort();
		}
	printf("\n");
	return ret;
}

int main(int argc, char *argv[])
{
	int ret = -1, dtaexists, idxexists, blkexists;
	mkoptions.create = 0;
	mkoptions.barrier = 0;
	mkoptions.commit_interval = 0;
	mkoptions.datafile = NULL;
	mkoptions.cachesize = 0;
	mkoptions.sectorsize = 0;
	struct stat stdta;
	struct stat device;
	int mode = O_RDWR;
	int fd;
	int dev;

	if (argc < 3)
		usage(argv[0]);
	if (0 != get_opts(argc, argv))
		exit(-1);

	if ((fd = open("/dev/eprdcontrol", mode)) < 0) {
		fprintf(stderr,
			"Failed to open /dev/eprdcontrol, is eprd.ko loaded?\n");
		exit(-1);
	}

	if (-1 == flock(fd, LOCK_EX)) {
		fprintf(stderr, "Failed to lock /dev/eprdcontrol\n");
		exit(-1);
	}

	if (mkoptions.device) {
		if (-1 == stat(mkoptions.device, &device)) {
			fprintf(stderr, "No such device : %s\n",
				mkoptions.device);
			exit(-1);
		}
		if (!S_ISBLK(device.st_mode)) {
			fprintf(stderr, "Not a blockdevice : %s\n",
				mkoptions.device);
			exit(-1);
		}
		ret = eprd_setup(EPRD_DEREGISTER, fd);
		if (0 != ret)
			die_ioctlerr("ioctl EPRD_DEREGISTER failed\n");
		exit(ret);
	}
	if (mkoptions.create)
		printf("creating a new blockdevice\n");
	else
		printf("Using an existing blockdevice\n");
	printf("datafile=%s\n", mkoptions.datafile);
	dtaexists = stat(mkoptions.datafile, &stdta);

	if (mkoptions.create && 0 == dtaexists) {
		fprintf(stderr,
			"Data file %s already exists, cowardly refusing to overwrite.\n",
			mkoptions.datafile);
		goto end_exit;
	}

	fprintf(stderr, "commit_interval = %u\n", mkoptions.commit_interval);

	if (mkoptions.create) {
		mkoptions.dtasize /= 4096;
		mkoptions.dtasize *= 4096;
		printf("Blockdevice size = %li\n", mkoptions.dtasize);
	} else {
		if (S_ISBLK(stdta.st_mode)) {
			if ((dev = open(mkoptions.datafile, O_RDONLY)) < 0) {
				fprintf(stderr, "Failed to open %s\n",
					mkoptions.datafile);
				exit(-1);
			}
			mkoptions.dtasize = lseek64(dev, 0, SEEK_END);
			if (-1 == mkoptions.dtasize) {
				fprintf(stderr, "Error while opening %s : %s\n",
					mkoptions.datafile, strerror(errno));
				goto end_exit;
			}
			mkoptions.dtasize /= 4096;
			mkoptions.dtasize *= 4096;
			close(dev);
		} else {
			mkoptions.dtasize = (unsigned long)stdta.st_size;
		}
		printf("Blockdevice size = %li\n", mkoptions.dtasize);
	}

	if (mkoptions.dtasize < 1048576) {
		fprintf(stderr, "Blockdevice %s with size %li is to small\n",
			mkoptions.datafile, mkoptions.dtasize);
		goto end_exit;
	}

	ret = eprd_setup(EPRD_INIT, fd);
	if (0 != ret)
		die_ioctlerr("ioctl EPRD_INIT failed\n");

	if (NULL != mkoptions.datafile) {
		ret = eprd_setup(EPRD_SET_DTAFD, fd);
		if (0 != ret)
			die_ioctlerr("ioctl EPRD_SET_DTAFD failed\n");
	}
	if (mkoptions.barrier) {
		ret = eprd_setup(EPRD_BARRIER, fd);
		if (0 != ret)
			die_ioctlerr("ioctl EPRD_BARRIER failed\n");
	}

	if (0 != mkoptions.cachesize) {
		printf("Cache size = %lu\n", mkoptions.cachesize);
		ret = eprd_setup(EPRD_CACHESIZE, fd);
		if (0 != ret)
			die_ioctlerr("ioctl EPRD_CACHESIZE failed\n");
	}

	if (0 != mkoptions.sectorsize) {
		printf("Sector size = %u\n", mkoptions.sectorsize);
		ret = eprd_setup(EPRD_SET_SECTORSIZE, fd);
		if (0 != ret)
			die_ioctlerr("ioctl EPRD_SET_SECTORSIZE failed\n");
	}

	ret = eprd_setup(EPRD_SET_COMMIT_INTERVAL, fd);
	if (0 != ret)
		die_ioctlerr("ioctl EPRD_SET_COMMIT_INTERVAL failed\n");

	ret = eprd_setup(EPRD_REGISTER, fd);
	if (0 != ret)
		die_ioctlerr("ioctl EPRD_REGISTER failed\n");
      end_exit:
	flock(fd, LOCK_UN);
	close(fd);
	exit(ret);
}
